### DISCUSSION TITLE
Precision Issue with printf Functions for Decimal Numbers specially subnormal values

### DISCUSSION THREAD

**Discussion by A-Shehab** [picolibc/picolibc#809] [Q&A]
Hello everyone,

I’ve encountered an issue with the printf function when printing numbers with a high number of decimal places. The output is not as accurate as expected. For instance.

```
#include <stdio.h>

int main()
{
    /* Test Case 1: Subnormal value to print */
    {
        volatile double x = 1.482196E-323;
        printf("x = %.16e\n", x);
    }
    /* Test Case 2: Random large value to print */
    {
       volatile long double x = 452123242.99999905L;
       printf("x = %.17e\n", x);
    }
    return 0;
}
```
Expected Behavior:
Test Case 1 prints:  = 1.4821969375237396e-323
Test Case 2 prints:  = 4.52123242999999046e+08

Observed Behavior:
Test Case 1 prints:  = 1.5000000000000000e-323
Test Case 2 prints:  = 4.52123242999999050e+08

with random large numbers such as test case 2 the difference is not huge but still it differs, but working with subnormal values the inaccuracy increases..

picolibc working with floating-point numbers is based on the ryu implementation, debugging the subnormal value for example when determining the valid decimal representation for this value at the first iteration:
v+=172
v=148
v-=123
so setting the upper and lower bounds were not accurate or tight enough resulting in a less precise output.

So my question is that a precision limitation working with this algorithm specially with subnormal values ? 

It would be helpful to understand if this is an issue or there any config in picolibc that could handle such cases more accurately.

any suggestions would be appreciated, thank you

**Comment by keith-packard** [comment]
For 1, the answer is that picolibc prints values with enough precision so that there is no ambiguity between successive floating point values. As the denorm value provided has only a few bits of significand remaining, two decimal digits is sufficient to produce a unique result. This makes the output precision mirror the precision of the binary representation. Generating more digits to produce a value which is closer to an infinite precision version of the binary value would require more precision in the conversion computation.

For 2, you should first note that picolibc doesn't (yet) have ryu code for values beyond 64 bits, but it appears from the answer you've gotten that you're using a target with 64-bit long double values as the %.17e works while targets with long double beyond 64 bits would require %.17le. Beyond that, again, picolibc uses the number of digits necessary to generate a result which unambgiuously represents the internal binary value. 17 digits is always sufficient for normal 64-bit floats, and so you get 17 digits with an extra zero tacked on because you wanted 18.

**Reply by tagunil** [reply]
Hello @keith-packard, and thanks for the detailed explanation.

I've also encountered the difference between picolibc and other libraries while printing subnormals. Although your explanation makes a lot of numerical sense, it would be really nice to have a "compatibility" mode, producing the results similar to what other C libraries do. I wonder how bad it would be in terms of performance and implementation complexity.

**Reply by keith-packard** [reply]
It would be extremely expensive -- all of those libraries perform arbitrary precision arithmetic for those operations, and that requires runtime allocation. With the ryu code, it's all fixed precision and thus requires no malloc at all.
